src/libraries/System.Web.HttpUtility/src/System/Web/Util/HttpEncoder.cs

```diff
@@ -611,15 +558,49 @@ private static string UrlPathEncodeImpl(string value)
                 return value;
             }
 
-            // recurse in case there is a query string
-            int i = value.IndexOf('?');
-            if (i >= 0)
+            int i = value.AsSpan().IndexOfAnyExceptInRange((char)0x21, (char)0x7F);
+            if (i < 0)
+            {
+                return value;
+            }
+
+            int indexOfQuery = value.IndexOf('?');
+            if (indexOfQuery >= 0 && indexOfQuery < i)
             {
-                return string.Concat(UrlPathEncodeImpl(value.Substring(0, i)), value.AsSpan(i));
+                // Everything before the Query is valid ASCII
+                return value;
+            }
+
+            ReadOnlySpan<char> toEncode = indexOfQuery >= 0
+                ? value.AsSpan(i, indexOfQuery - i)
+                : value.AsSpan(i);
+
+            byte[] bytes = ArrayPool<byte>.Shared.Rent(Encoding.UTF8.GetMaxByteCount(toEncode.Length));
+            char[] chars = ArrayPool<char>.Shared.Rent(bytes.Length * 3);
+
+            int utf8Length = Encoding.UTF8.GetBytes(toEncode, bytes);
+            int charCount = 0;
+            foreach (byte b in bytes.AsSpan(0, utf8Length))
+            {
+                if (IsNonAsciiOrSpaceByte(b))
+                {
+                    chars[charCount++] = '%';
+                    chars[charCount++] = HexConverter.ToCharLower(b >> 4);
+                    chars[charCount++] = HexConverter.ToCharLower(b);
+                }
```

Comment:
To avoid some bound-checks is
```suggestion
                {
                    Span<char> span = chars.AsSpan(charCount);
                    span[2] = HexConverter.ToCharLower(b);
                    span[1] = HexConverter.ToCharLower(b >> 4);
                    span[0] = '%';
                }
```
faster?

I don't know if current JIT needs the reverse indexes (bound-check only on `[2]`, then it knows that the other indexes are safe) or if the normal ascending indexes also work with this.

Maybe also move the creation of the span above the loop, then slice it only inside the loop.

Reply:
@MihaZupan any objection against these changes? They are a bit not intuitive to my taste.

Reply:
I don't think it's that important for this particular method.

But the pattern does appear in a bunch of places, so I would prefer we add a shared helper for this instead of optimizing each call site.
A few examples:
https://github.com/dotnet/runtime/blob/59f2833b98e15e0831be8ad24fb321ee7bf9bd54/src/libraries/System.Private.Uri/src/System/Uri.cs#L1383-L1384
https://github.com/dotnet/runtime/blob/59f2833b98e15e0831be8ad24fb321ee7bf9bd54/src/libraries/System.Private.Uri/src/System/UriHelper.cs#L511-L515
https://github.com/dotnet/runtime/blob/59f2833b98e15e0831be8ad24fb321ee7bf9bd54/src/libraries/System.Private.CoreLib/src/System/Net/WebUtility.cs#L356-L358
https://github.com/dotnet/runtime/blob/59f2833b98e15e0831be8ad24fb321ee7bf9bd54/src/libraries/System.Net.Http/src/System/Net/Http/Headers/HeaderUtilities.cs#L115-L120

Reply:
(I wouldn't do it as part of this PR though)